--- a/nat46/modules/nat46-core.c	2020-08-20 20:28:43.000000000 +0530
+++ b/nat46/modules/nat46-core.c	2023-09-29 12:52:34.699230218 +0530
@@ -1798,7 +1798,7 @@
   }
 
   if(!pairs_xlate_v4_to_v6_outer(nat46, hdr4, having_l4 ? &sport : NULL, having_l4 ? &dport : NULL, v6saddr, v6daddr)) {
-    nat46debug(0, "[nat46] Could not translate v4->v6");
+    //nat46debug(0, "[nat46] Could not translate v4->v6");
     goto done;
   }
 
